### DISCUSSION TITLE
What's the best practice to save and load trained/fine-tuned models e.g. `BertClassifier`?

### DISCUSSION THREAD

**Discussion by alvarobartt** [keras-team/keras-hub#1039] [Q&A]
Hi @jbischof! I was wondering what's the best way to save and load any `...Classifier` model e.g. `BertClassifier` that has been trained/fine-tuned via `.fit` on custom data.

I assume that it should be `.save` and then `tf.keras.models.load_model`, but the latter is complaining due to missing `inputs` and `outputs` (`TypeError: Functional.__init__() missing 2 required positional arguments: 'inputs' and 'outputs'`).

So I was wondering whether there's a clear approach on how to properly save and load models. I saw some discussions about some proposals of HuggingFace-like interfaces, but not sure if there's been any clear decision on that.

Thanks!

**Comment by alvarobartt** [comment]
P.S. for more context, I was experimenting with `BertClassifier.from_preset("bert_base_en", num_classes=2)` to fine-tune it for GLUE SST-2 for experimentation purposes. And looking not just for a solution on how to load any fine-tuned model, but also on how to do that easily (user-friendly speaking).

**Reply by alvarobartt** [reply]
So this being said, is there any way to train a model like:

```python
import keras_nlp
classifier = keras_nlp.models.BertClassifier.from_preset("bert_tiny_en_uncased", num_classes=2)
classifier.backbone.trainable = False
classifier.fit(custom_train_ds, validation_data=custom_val_ds, epochs=10)
classifier.save("MODEL")
```

And then load it in an easier way than:

```
fine_tuned_classifier = keras_nlp.models.BertClassifier.from_preset("bert_tiny_en_uncased", num_classes=2, load_weights=False)
fine_tuned_classifier.load_weights("MODEL")
```

That would help avoid the definition of `num_classes` since shouldn't need to be known by the user that wants to use a fine-tuned model, and also having to load the weights separately. So that one could do e.g. `keras_nlp.models.BertClassifier.from_pretrained("MODEL")` and that's it.

**Reply by alvarobartt** [reply]
I know this has been widely discussed at https://github.com/keras-team/keras-nlp/discussions/355, and you submitted a proposal at https://github.com/keras-team/keras-nlp/issues/377

**Reply by alvarobartt** [reply]
Maybe a thin wrapper on top of `.save` could do the work to dump a configuration file too, similar to the presets, in the same directory as the checkpoints, so that just the path to the saved model had to be specified instead via e.g. `from_checkpoint`, `from_pretrained`, or any other option

**Reply by jbischof** [reply]
@alvarobartt saving should work out of the box (the issues you linked are old API design discussions).

If you are hitting a bug my suggestion is to file an issue with a minimal colab required to reproduce and we will triage. Thank you!

**Reply by alvarobartt** [reply]
@jbischof Yes, both saving and loading work fine, I was just wondering whether there's an easier way to load a fine-tuned model back from e.g. `BertClassifier` rather than having to instantiate it first with the base model name and the number of classes, and then loading the weights separately.

**Reply by alvarobartt** [reply]
Since for the moment you have `from_preset` which does that but only for the available models, there's not an easy way to e.g. register a new preset, or load a fine-tuned model using the same preset schema, etc.

**Reply by jbischof** [reply]
`model.save()` should serialize the weights along with the architecture. What's the feature gap?

presets are an implementation detail of the library....unless we go from a curated library to full 🤗

**Reply by alvarobartt** [reply]
Yes, it does, but one's not able to `keras.models.load_model("<MODEL_PATH>")`, as it's expecting both inputs and outputs too, so there's no way to load a fine-tuned model with the path only, you need to know more details in advance, so there's not a clear way on how to do that in an abstract way, so that anyone can load any model just with the model path (to the `SavedModel` files).

**Reply by jbischof** [reply]
That sounds like a bug. Could you please file an issue?

See https://keras.io/guides/serialization_and_saving/ as a general guide for how saving should work.

**Reply by alvarobartt** [reply]
Sure @jbischof, creating it rn!
